Convert gameplay leaderboard to skinnable component

bdach opened the pull request:
Makes the leaderboard a full skinnable component (placed in the ruleset-specific layer).

Users with skins that have custom HUD layouts need to add the gameplay leaderboard back themselves in their desired position.

This change also allows the leaderboard to stop overlapping the playfield in taiko.

| | argon | triangles | classic |
| :-: | :-: | :-: | :-: |
| osu! | ![Screenshot 2025-04-24 at 13 02 58](https://github.com/user-attachments/assets/7c3c03ba-ab88-4f4d-8e85-a9eb050cd5b4) | ![Screenshot 2025-04-24 at 13 10 56](https://github.com/user-attachments/assets/f9147758-9363-4fcb-9c19-8af8ea3d2c6c) | ![Screenshot 2025-04-24 at 13 03 41](https://github.com/user-attachments/assets/1dd477dc-b6e8-464e-9f41-264a388c1e50) |
| taiko | ![Screenshot 2025-04-24 at 13 03 06](https://github.com/user-attachments/assets/a58cc95e-d47a-4f7e-b81e-bdf1b909fdfc) | ![Screenshot 2025-04-24 at 13 09 03](https://github.com/user-attachments/assets/dbfd3bd1-d6e7-4675-9d6e-14ea4f40f058) | ![Screenshot 2025-04-24 at 13 03 47](https://github.com/user-attachments/assets/ff7b81b7-97e8-44f2-ba2a-c5ef26cad3a6) |
| catch | ![Screenshot 2025-04-24 at 13 03 11](https://github.com/user-attachments/assets/572804ff-9f67-48e1-8d30-bb4ab25b28bc) | ![Screenshot 2025-04-24 at 13 10 59](https://github.com/user-attachments/assets/bb944b3b-3e34-478b-94ba-cb8854074c41) | ![Screenshot 2025-04-24 at 13 03 53](https://github.com/user-attachments/assets/d6afb6a9-f404-4d9a-9b46-5b5621e54302) |
| mania | ![Screenshot 2025-04-24 at 13 03 15](https://github.com/user-attachments/assets/913e67e7-da27-4e5f-be77-6e08e0239962) | ![Screenshot 2025-04-24 at 13 11 03](https://github.com/user-attachments/assets/df95a06f-ce42-402d-abfa-d56509cb66ee) | ![Screenshot 2025-04-24 at 13 03 58](https://github.com/user-attachments/assets/2dca8c30-d31d-4d43-88b5-6ed1638dc915) |

---

- [x] Depends on https://github.com/ppy/osu/pull/32938 for LoC reduction

This PR converts the leaderboard into a full-fledged skinnable component that can be manipulated by users at will.

Notably, this finally allows https://github.com/ppy/osu/issues/20422 to be fixed - although it's a very mixed bag, for several reasons:

- Because of taiko players' ~~refusal to see reason~~ insistence on   keeping stable behaviours related to aspect ratio treatment, I have to assume the worst case scenario, which means than on typical resolutions like 16:9 (or even worse, 4:3), the leaderboard will likely not occupy as much vertical space as it probably could.
- Additionally, there's the problem of where to put the spectator list. I settled on putting it to the right of the leaderboard, but that's kind of janky, because the leaderboard sometimes collapses and sometimes fully hides, leading to a very awkward space left behind. If we had the capability to anchor elements to other elements, maybe this could be resolved, but for now, I'm not sure what to do. I think if some users are bothered by it they can move it where they want it. Or delete it.

As with the spectator list, if some users have modified the layout on their skin, they will not have the leaderboard added back, and I'm not trying to add any migrations to try and do it for them.

Recommend reviewing with whitespace off.

bdach committed:
Convert gameplay leaderboard to skinnable component

This PR converts the leaderboard into a full-fledged skinnable component
that can be manipulated by users at will.

Notably, this finally allows https://github.com/ppy/osu/issues/20422 to
be fixed - although it's a very mixed bag, for several reasons:

- Because of taiko players' refusal to see reason^W^W^W^Winsistence on
  keeping stable behaviours related to aspect ratio treatment, I have to
  assume the worst case scenario, which means than on typical
  resolutions like 16:9 (or even worse, 4:3), the leaderboard will
  likely not occupy as much vertical space as it probably could.

- Additionally, there's the problem of where to put the spectator list.
  I settled on putting it to the right of the leaderboard, but that's
  kind of janky, because the leaderboard sometimes collapses and
  sometimes fully hides, leading to a very awkward space left behind. If
  we had the capability to anchor elements to other elements, maybe this
  could be resolved, but for now, I'm not sure what to do. I think if
  some users are bothered by it they can move it where they want it. Or
  delete it.

bdach commented on osu.Game/Screens/Play/HUD/DrawableGameplayLeaderboard.cs:
Depending on whether we want to offer a "legacy" variant of the leaderboard matching stable skins, it might be a good idea to rename this component to something like `ArgonGameplayLeaderboard` before it gets persisted to the JSONs.

peppy replied:
I'm fine with either. We already have a good flow for migrating component names so it's not a huge deal.

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerPlayer.cs:
```diff
@@ -71,10 +70,7 @@ private void load()
 
-            LoadComponentAsync(new GameplayChatDisplay(Room)
-            {
-                Expanded = { BindTarget = LeaderboardExpandedState },
```
This being coupled to leaderboard expansion state is a bit of a spoke in the wheel. I simplified this to just use `LocalUserPlaying` which is one half of what was steering the expansion state. Maybe fine?

peppy replied:
I think it's fine, yeah.

peppy replied:
Or maybe not?

![osu! 2025-05-07 at 09 36 04](https://github.com/user-attachments/assets/8d48c58c-e5f8-49e4-bb42-0b937137a8dc)

Doesn't stay expanded in `MultiplayerSpectator` anymore.

bdach committed:
Fix code quality

bdach commented on osu.Game.Rulesets.Taiko/Skinning/Legacy/TaikoLegacySkinTransformer.cs:
```diff
@@ -29,119 +32,180 @@ public TaikoLegacySkinTransformer(ISkin skin)
+                                    leaderboard.Anchor = Anchor.BottomLeft;
+                                    leaderboard.Origin = Anchor.BottomLeft;
+                                    leaderboard.Position = pos;
+                                    leaderboard.Height = 170;
```
This height can be improved later when the positioning of taiko combo counter is adjusted to match stable (should be on the drum).

bdach committed:
Fix the same test that I forget to update every time

bdach committed:
Delete no-longer-relevant test

bdach committed:
Fix one more test failure

bdach committed:
Fix some weirdness around forced multiplayer elements

- Naming wasn't adjusted
- When the chat collapsed, the team score display broke due to zero
  width

Partial revert of
https://github.com/ppy/osu/commit/da1fc1013e07b8dafb0c409354f9d1cef971e449.

bdach committed:
Yet another forgotten initialiser

bdach committed:
Merge branch 'master' into gameplay-leaderboard-skinnable

bdach committed:
Merge branch 'master' into gameplay-leaderboard-skinnable

smoogipoo reviewed (dismissed):
Looks ok to me

peppy committed:
Merge branch 'master' into gameplay-leaderboard-skinnable

peppy reviewed (commented):
See https://github.com/ppy/osu/pull/32939#discussion_r2077237972

bdach committed:
Fix gameplay leaderboard not always being expanded in gameplay leaderboard

I'd have preferred a `get; init;` property but tests were also attached
at the hip to the public bindable. Without some extra composition this
is the best that I can do.

bdach committed:
Fix back-to-front conditional

bdach commented:
> See [#32939 (comment)](https://github.com/ppy/osu/pull/32939#discussion_r2077237972)

Ugh. Extremely annoying oversight. Should be fixed with https://github.com/ppy/osu/pull/32939/commits/8cc2af4060bfb8a9c1f7ede893aae75bfb219e2e. I would have preferred that to not be bindable but unfortunately a bunch of tests also want a public bindable. This was the path of least resistance (without introducing some composition contraption wherein the drawable leaderboard just has `Expanded` but not hooked up to any of the configs, and then having a "gameplay" variant that is the actual skinnable and also wraps it and plugs into the config appropriately). Not sure what to think of it.

Also noticed https://github.com/ppy/osu/pull/32939/commits/1fc68a3c485f090998595bfe1a18605b22515412 along the way. Oops? 🤦
